text/0000-macro-naming.md

```diff
@@ -0,0 +1,169 @@
+
+E.g., `#[::foo::bar::baz(...)]` and `#[bar::baz(...)]` are uses of absolute and
+relative paths, respectively.
+
+
+## Importing macros
+
+Importing macros is done using `use` in the same way as other items. An `!` is
+not necessary in an import item. Macros are imported into their own namespace
+and do not shadow or overlap items with the same name in the type or value
+namespaces.
+
+E.g., `use foo::bar::baz;` imports the macro `baz` from the module `::foo::bar`.
+Macro imports may be used in import lists (with other macro imports and with
+non-macro imports).
+
+Where a glob import (`use ...::*;`) imports names from a module including macro
+definitions, the names of those macros are also imported. E.g., `use
+foo::bar::*;` would import `baz` along with any other items in `foo::bar`.
+
+Where macros are defined in a separate crate, these are imported in the same way
+as other items by an `extern crate` item.
+
+No `#[macro_use]` or `#[macro_export]` annotations are required.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If the new macro system is not well adopted by users, we could be left with two
+very different schemes for naming macros depending on whether a macro is defined
+by example or procedurally. That would be inconsistent and annoying. However, I
+hope we can make the new macro system appealing enough and close enough to the
+existing system that migration is both desirable and easy.
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+We could adopt the proposed scheme for procedural macros only and keep the
+existing scheme for macros by example.
+
+We could adapt the current macros by example scheme to procedural macros.
+
+We could require the `!` in macro imports to distinguish them from other names.
+I don't think this is necessary or helpful.
+
+We could continue to require `macro_export` annotations on top of this scheme.
+However, I prefer moving to a scheme using the same privacy system as the rest
+of Rust, see below.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Privacy for macros
+
+I would like that macros follow the same rules for privacy as other Rust items,
+i.e., they are private by default and may be marked as `pub` to make them
+public. This is not as straightforward as it sounds as it requires parsing `pub
+macro! foo` as a macro definition, etc. I leave this for a separate RFC.
+
+## Scoped attributes
+
+It would be nice for tools to use scoped attributes as well as procedural
+macros, e.g., `#[rustfmt::skip]` or `#[rust::new_attribute]`. I believe this
+should be straightforward syntactically, but there are open questions around
+when attributes are ignored or seen by tools and the compiler. Again, I leave it
+for a future RFC.
+
+## Inline procedural macros
+
+Some day, I hope that procedural macros may be defined in the same crate in
+which they are used. I leave the details of this for later, however, I don't
+think this affects the design of naming - it should all Just Work.
```

Comment:
Procedural macros and any items marked with `#[cfg(macro)]` would be split off and compiled into one crate, everything else compiled into another. The first crate would be linked in to the compiler while it compiles the second (or equivalently, called using IPC).